### PR TITLE
Make `{Arc,Rc,Weak}::ptr_eq` ignore pointer metadata

### LINKED ISSUE #103763: `Arc::ptr_eq` does not always return "true if the two Arcs point to the same allocation" as documented
# How `Arc::ptr_eq` is documented, and current expectations:

For reference here this is the (I think concise) documentation for `Arc::ptr_eq`:
> Returns true if the two Arcs point to the same allocation (in a vein similar to [ptr::eq](https://doc.rust-lang.org/std/ptr/fn.eq.html)).

My expectation from this is rather simple: the implementation will get the pointer for the `Box`/"allocation" that is allocated as the backing storage for two arcs (holding the reference count and inner value) and it will compare those pointers with `ptr::eq`.

_Notably: the wrapped type should have no bearing on the implementation / semantics here, since it's the pointer of the "allocation" that is supposed (documented) to be compared._

This way you can create any number of references for a single `Arc` that wraps some inner value and cheaply determine whether two `Arc`s point back to the same state allocation.

I've been relying on those semantics for `Arc::ptr_eq` across a number of projects for _exactly_ that purpose.

For a few projects I should also note that I have generalized how I wrap state in an `Arc` as an aid for also exposing that state across FFI boundaries. I note this just to highlight the separation of concerns that I have relied on - namely that I assume I can use `Arc::ptr_eq` to check whether two Arcs point to the same state without having to worry about what kind of type is wrapped by the Arc.

# Actual implementation of `Arc::ptr_eq`:

Instead of comparing the pointers of the backing "allocation", the current implementation of `Arc::ptr_eq` instead figures out the offset pointer to the inner _value_ and does a comparison of those value pointers with semantics that could vary depending on what is wrapped by the Arc.

I.e. it does `Arc::as_ptr(&a) == Arc::as_ptr(&b)`

In particular if an Arc contains a `dyn Trait` then the pointers would be wide/fat pointers and the comparison will compare the vtable pointer for those trait objects which could be equal even though the two `Arc`s are otherwise unrelated, not referencing the same "allocation". (I.e. it could return `true` for two Arcs that **don't** point to the same allocation)

Especially in the context of state that's being shared over unsafe FFI boundaries I think this deviation from the documentation could be rather dangerous.

Even unrelated to FFI this breaks the basic notion that you can rely on this API compare whether two Arcs reference the same state which may easily break logical invariants that code may rely on, for example to implement the `Eq` trait of a higher level type.

As it happens I discovered this in a project for Json Web Tokens after I introduced Clippy to the project's CI. There was some validation state that was wrapped in an `Arc` (closures that the application could add for validating a token) and discovering that the logic for comparing this validation state is broken by this was particularly concerning for me.

# Duplicate issue

This issue was actually raised before as https://github.com/rust-lang/rust/pull/80505 but I think it was a mistake that the discussion focused on the question of consistency between `ptr::eq` and `Arc::ptr_eq` in terms of how wide pointers should be compared.

The documentation for `Arc::ptr_eq` gives no suggestion that it is comparing pointers to the wrapped type and so there is no cause for concern regarding the consistent handling of wide pointer comparisons.

Even considering the mention of `ptr::eq` in the documentation where it says:
> (in a vein similar to [ptr::eq](https://doc.rust-lang.org/std/ptr/fn.eq.html))

there is no implication that the inner type of the `Arc` affects what pointers are being compared.

`ptr::eq` should conceptually be used to compare the pointers to the `Arc`s inner allocation, which should just be the raw (non-wide) pointer for the `Box` that encompasses the `Arc`s reference count. In this way the comparison is done "in a vein similar to `ptr::eq`" (it may literally use `ptr::eq` internally).

# Previous agreement that this is a bug

The discussion for #80505 seemed to go around in circles but it seems important to highlight that the libraries team did discuss the issue and actually agreed that it was a bug https://github.com/rust-lang/rust/pull/80505#issuecomment-757141449 

With that agreement it's hard to understand why #80505 was closed.

### Meta

`rustc --version --verbose`:
```
rustc 1.64.0 (a55dd71d5 2022-09-19)
binary: rustc
commit-hash: a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52
commit-date: 2022-09-19
host: x86_64-pc-windows-msvc
release: 1.64.0
LLVM version: 14.0.6
```

### PR DESCRIPTION
**Pull request by albertlarsan68**
FCP completed in https://github.com/rust-lang/rust/issues/103763#issuecomment-1362267967

Closes #103763

### CODE REVIEW

File: src/tools/clippy/clippy_lints/src/unnamed_address.rs [hunk 1]
```diff
@@ -96,9 +96,7 @@ impl LateLintPass<'_> for UnnamedAddress {
             if let ExprKind::Call(func, [ref _left, ref _right]) = expr.kind;
             if let ExprKind::Path(ref func_qpath) = func.kind;
             if let Some(def_id) = cx.qpath_res(func_qpath, func.hir_id).opt_def_id();
-            if match_def_path(cx, def_id, &paths::PTR_EQ) ||
-                match_def_path(cx, def_id, &paths::RC_PTR_EQ) ||
-                match_def_path(cx, def_id, &paths::ARC_PTR_EQ);
```
Comment (flip1995) [inline src/tools/clippy/clippy_lints/src/unnamed_address.rs:101]: AFAICT those paths were only used here. Could you also remove them from: 
[clippy_utils/src/paths.rs](https://github.com/rust-lang/rust-clippy/blob/0bca8dd2547fcc46796b53d338c0734575e51c72/clippy_utils/src/paths.rs#L18), please.

Comment (albertlarsan68) [reply src/tools/clippy/clippy_lints/src/unnamed_address.rs:101]: Done !

File: library/alloc/src/rc.rs [hunk 2]
```diff
@@ -1153,7 +1153,7 @@ impl<T: ?Sized> Rc<T> {
     #[inline]
     #[stable(feature = "ptr_eq", since = "1.17.0")]
     /// Returns `true` if the two `Rc`s point to the same allocation in a vein similar to
-    /// [`ptr::eq`]. See [that function][`ptr::eq`] for caveats when comparing `dyn Trait` pointers.
+    /// [`ptr::eq`]. Here, `dyn Trait` pointers' metadata are ignored.
```
Comment (pitaj) [inline library/alloc/src/rc.rs:1156]: This phrasing feels clunky to me. Maybe this instead:

```suggestion
    /// [`ptr::eq`]. However, this function ignores the metadata of `dyn Trait` pointers.
```

Comment (albertlarsan68) [reply library/alloc/src/rc.rs]: Changed!

File: library/alloc/src/rc.rs [hunk 3]
```diff
@@ -1153,7 +1153,7 @@ impl<T: ?Sized> Rc<T> {
     #[inline]
     #[stable(feature = "ptr_eq", since = "1.17.0")]
     /// Returns `true` if the two `Rc`s point to the same allocation in a vein similar to
-    /// [`ptr::eq`]. See [that function][`ptr::eq`] for caveats when comparing `dyn Trait` pointers.
+    /// [`ptr::eq`]. However, this function ignores the metadata of  `dyn Trait` pointers.
```
Comment (Amanieu) [inline library/alloc/src/rc.rs:1156]: Slight language nit: the "However" should be removed. Simply saying "This function ignores the metadata of ..." is enough.

File: library/alloc/src/rc.rs [hunk 4]
```diff
@@ -1168,7 +1168,7 @@ impl<T: ?Sized> Rc<T> {
     /// assert!(!Rc::ptr_eq(&five, &other_five));
     /// ```
     pub fn ptr_eq(this: &Self, other: &Self) -> bool {
-        this.ptr.as_ptr() == other.ptr.as_ptr()
+        ptr::eq(this.ptr.as_ptr() as *const (), other.ptr.as_ptr() as *const ())
```
Comment (Amanieu) [inline library/alloc/src/rc.rs:1171]: You can use `==` here, you don't need to call `ptr::eq`. The cast is the important part.

**Comment by albertlarsan68** [comment]
@m-ou-se This PR is ready to be reviewed.

**Comment by albertlarsan68** [comment]
Thanks Amanieu!

**Pull request merged**